sky/packages/sky/lib/src/services/shell.dart

```diff
@@ -18,15 +18,38 @@ ApplicationConnection _initConnection() {
   return new ApplicationConnection(null, serviceProvider);
 }
 
+// A function that takes a Mojo service URL and returns an implementation of
+// that service's interface.
+typedef Object OverrideRequestService(String url);
+
+// Set this to intercept calls to requestService and supply an alternative
+// implementation of a service (for example, a mock for testing).
+OverrideRequestService overrideRequestService;
+
 class _ShellImpl {
   _ShellImpl._();
 
   final ApplicationConnection _connection = _initConnection();
 
-  void requestService(String url, Object proxy) {
+  void _requestService(String url, Object proxy) {
     if (embedder.shell == null) _connection.requestService(proxy);
     else embedder.connectToService(url, proxy);
   }
+
+  void requestService(String url, Object proxy) {
+    if (overrideRequestService == null) {
+      _requestService(url, proxy);
+    } else {
+      Object overrideImpl = overrideRequestService(url);
+      if (overrideImpl != null) {
+        // Replace the proxy's implementation of the service interface with the
+        // one provided by the override.
+        proxy.ptr = overrideImpl;
```

Comment:
This seems a bit too opinionated for this file.  Can we move this logic into a library in the tests and have a more general delegation point here?